Bio/Align/_pairwisealigner.c

```diff
@@ -3796,13 +3802,81 @@ Aligner_get_algorithm(Aligner* self, void* closure)
                     break;
             }
             break;
+        case FOGSAA:
+            switch (mode) {
+                case Global:
+                    s = "Fast Optimal Global Sequence Alignment Algorithm";
+                    break;
+                case Local:
+                    PyErr_SetString(PyExc_RuntimeError,
+                            "algorithm does not support local mode");
+                    return NULL;
+            }
         case Unknown:
         default:
             break;
     }
     return PyUnicode_FromString(s);
 }
 
+static int
+Aligner_set_algorithm(Aligner* self, PyObject* value, void* closure)
+{
+    if (value == Py_None) {
```

Comment:
I haven't tried the code, but it seems that with this method the user is able to choose an algorithm that is inconsistent with the gap scores.

Reply:
Good catch. I pushed another commit today that should fix this. Can you take a look at it?

Reply:
I think that the algorithm should be read-only. The appropriate algorithm is chosen automatically based on the gap scores and mode. Other than for FOGSAA, there is no point in choosing a different algorithm, because then the algorithm is either inconsistent with the gap scores, or it is slower than the automatically selected one (while producing identical results).

Also, if a user sets the `algorithm` to Gotoh, but then changes the gap scores, the algorithm may flip back to Needleman-Wunsch if the gap scores are consistent with that algorithm.

I would suggest to make 'FOGSAA' (or 'fast global' or whatever you want to call it) one of the options for the `mode` attribute. If that option is chosen, the `algorithm` attribute automatically switches to `FOGSAA`.